### PR TITLE
chaos-dashboard web api for workflow

### PR DESCRIPTION
**Pull request by STRRL**
### What problem does this PR solve?

Implementation of chaos-dashboard web API, should not be merged before #1551.

### What is changed and how does it work?
Implementation of chaos-dashboard web API, build a repo with kube client.
And it updates the API about the parameter about list, get, and the entity of Workflow and WorkflowDetail

### Checklist

Tests

- [x] Unit test
- [ ] E2E test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```
Signed-off-by: STRRL <str_ruiling@outlook.com>

### CODE REVIEW

**Comment by STRRL** [comment]
PTAL 
/cc @g1eny0ung

File: pkg/core/workflow.go [hunk 1]
```diff
@@ -97,49 +103,12 @@ const (
 	TaskNode NodeType = "TaskNode"
 )
 
-// NodeState represents a node in different stage.
```
Comment (AsterNighT) [inline pkg/core/workflow.go:100]: Umm why are these statuses deleted?

**Review by AsterNighT — APPROVED** [review]

File: pkg/apiserver/workflow/workflow.go [hunk 2]
```diff
@@ -29,46 +32,76 @@ func Register(r *gin.RouterGroup, s *Service) {
 	endpoint := r.Group("/workflows")
 	endpoint.GET("", s.listWorkflows)
 	endpoint.POST("/new", s.createWorkflow)
-	endpoint.GET("/detail/:uid", s.getWorkflowDetail)
-	endpoint.DELETE("/:uid", s.deleteWorkflow)
-	endpoint.PUT("/:uid", s.updateWorkflow)
+	endpoint.GET("/detail/:namespace/:name", s.getWorkflowDetail)
```
[the remark this answers is not in the document]
Comment (g1eny0ung) [reply pkg/apiserver/workflow/workflow.go:35]: More simpler path, `GET /:namespace/:name`?

File: pkg/core/workflow.go [hunk 3]
```diff
@@ -154,3 +123,162 @@ type Template struct {
...
+		Name:      name,
+	}, &kubeWorkflow)
+	if err != nil {
+		return err
+	}
+	return it.kubeclient.Delete(ctx, &kubeWorkflow)
+}
+
+// func MutateWithKubeClient could spawn a new repo with the new kubeclient with another auth session.
+func (it *KubeWorkflowRepository) MutateWithKubeClient(anotherKubeclient client.Client) *KubeWorkflowRepository {
```
Comment (g1eny0ung) [inline pkg/core/workflow.go:203]: It seems that this method is no longer used.

File: pkg/apiserver/server.go [hunk 4]
```diff
@@ -37,6 +38,7 @@ var (
 		fx.Provide(
 			newEngine,
 			newAPIRouter,
+			core.NewKubeWorkflowRepository,
```
Comment (g1eny0ung) [inline pkg/apiserver/server.go:41]: We haven't provided the `client.Client` instance in `fx.provide`. I think this registration will fail.

Comment (g1eny0ung) [reply pkg/apiserver/server.go:41]: BTW, all APIs use `github.com/chaos-mesh/chaos-mesh/pkg/clientpool` as the kube client. Should workflow API also use it?

File: pkg/core/workflow.go (hunk 1 above)
Comment (STRRL) [reply pkg/core/workflow.go:100]: We have use Conditions to describe that in kubernetes API, I think we have no necessary to translate it to these NodeState(s).

File: pkg/apiserver/server.go (hunk 4 above)
Comment (STRRL) [reply pkg/apiserver/server.go:41]: yeah, we should use the `clientpool`, I will fix it ASAP.😅

**Comment by STRRL** [comment]
fixed @g1eny0ung , PTAL

/lgtm cancel

File: pkg/apiserver/server.go (hunk 4 above)
Comment (g1eny0ung) [reply pkg/apiserver/server.go:41]: Seems we don't need to provide this currently.

File: pkg/core/common.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+// KubeObjectYAMLDescription defines the YAML structure of an object stored in kubernetes API.
+type KubeObjectYAMLDescription struct {
+	APIVersion string                 `json:"apiVersion"`
+	Kind       string                 `json:"kind"`
+	Metadata   KubeObjectYAMLMetadata `json:"metadata"`
+	Spec       interface{}            `json:"spec"`
+}
+
+// KubeObjectYAMLMetadata defines the metadata of YAMLDescription.
```
Comment (g1eny0ung) [inline pkg/core/common.go:27]: `YAMLDescription` => `KubeObjectYAMLDescription`

Comment (STRRL) [reply pkg/core/common.go]: Fixed in f9459d6, PTAL

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Pull request merged**